Try box shadow, border, and notice for a template part editing indicators.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
This explores adding a semi-transparent box-shadow to a template part in the editor when it is selected for editing.

This is an alternative to the suggested "Spotlight Mode" behavior for template part editing.  Since that mode is already implemented on the individual block level, a box shadow may apply a similar style that is still distinctly different from and plays well with the spotlight mode.

While slightly rough and lacking on its own, this shadow may work well with other indicators that are being explored such as the renaming toolbar.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

### Before:
![Screen Shot 2020-06-23 at 6 22 12 PM](https://user-images.githubusercontent.com/28742426/85471507-8b8cf900-b57e-11ea-886d-d0eeeec777f6.png)

### After:
<img width="1255" alt="Screen Shot 2020-08-07 at 4 46 55 PM" src="https://user-images.githubusercontent.com/6265975/89696727-09f2ef80-d8ce-11ea-9cf1-3839fd5f4293.png">
1. Information about the change being persistent
2. Border visible when editing inner blocks
3. Overlay/scrim over rest of content when editing template part

![2020-08-07 16 45 59](https://user-images.githubusercontent.com/6265975/89696863-8ab1eb80-d8ce-11ea-8063-4f44169fb345.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
Add more visibility to template part

Addison-Stavlo commented:
Note - This idea would most likely require a function to determine the color/opacity of the shadow based on the current background color in the editor.  🤔   Otherwise this idea really won't work well for a variety of themes.

noahtallen commented:
I feel like this kind of thing might be "good enough" to start iterating on 🤔 Maybe also with a border or something

Addison-Stavlo commented:
> I feel like this kind of thing might be "good enough" to start iterating on 🤔 Maybe also with a border or something

Agree on both.  I pulled in changes from master now that it has width alignments, a renaming toolbar, and a border when the actualy template part block is selected (although we may want to have that border whenever inner blocks are selected as well?).

Template Part selected:
![Screen Shot 2020-07-16 at 6 40 05 PM](https://user-images.githubusercontent.com/28742426/87729712-facbc680-c793-11ea-8fbd-ab1c1a03e3af.png)

Inner block selected:
![Screen Shot 2020-07-16 at 6 40 14 PM](https://user-images.githubusercontent.com/28742426/87729732-0dde9680-c794-11ea-8c34-dca4d3e2024f.png)

At first glance Im not exactly sure where that border is being applied, but it may look nice to have it applied with inner blocks selected as well.

noahtallen commented:
Tried this out locally. Feels pretty nice, thanks for working on it!! I have some thoughts:

- the blue focus border definitely feels a little bit off. I'm so used to only seeing it when I have the block selected that it feels weird when I am working with other blocks. Maybe the border needs to be separate from the "active selection" state?
- There's a little issue with hover mode where it thinks the block is not hovered when going over whitespace in between child blocks:
![2020-08-06 14 54 25](https://user-images.githubusercontent.com/6265975/89586789-2ec76400-d7f5-11ea-8747-12d94063991e.gif)
- I think after seeing this, I like the notice UI closer to the designs where it is attached to the template part itself. As is, I actually didn't see it at first. (And IMO, I think I'd prefer it being always visible rather than going away after some amount of time 🤔 
<img width="1881" alt="Screen Shot 2020-08-06 at 2 54 47 PM" src="https://user-images.githubusercontent.com/6265975/89586926-66361080-d7f5-11ea-90b1-d35dc89aa5e7.png">
- I definitely like the "spotlight" around the template part when editing it.
- Can we also apply these changes to the post-content block? IMO, it fits the same pattern really well.

Addison-Stavlo commented:
> the blue focus border definitely feels a little bit off. I'm so used to only seeing it when I have the block selected that it feels weird when I am working with other blocks. Maybe the border needs to be separate from the "active selection" state?

By default the blue border is applied when the block is selected.  The changes added here make it also applied when the children in the template part are selected, is this what feels off?  i.e. the template part shouldn't have a blue border when we have selected one of its inner blocks?  I feel like the box shadow looked really naked without the border, but it will be simpler to leave it 'as-is' and may feel less naked if we are also adding the spotlight mode to the box shadow. 

> There's a little issue with hover mode where it thinks the block is not hovered when going over whitespace in between child blocks:

Yeah, it seems like the block inserter there is picking up the hover state.  I not sure how to get around this. 🤔

 > I think after seeing this, I like the notice UI closer to the designs where it is attached to the template part itself. As is, I actually didn't see it at first. (And IMO, I think I'd prefer it being always visible rather than going away after some amount of time

I think currently the notice options are to have the notice either at the top of the editor or on the snackbar.  With the snackbar it seems to disappear after some time by default, I wonder if there is an option to disable this behavior.

> I definitely like the "spotlight" around the template part when editing it. - Can we also apply these changes to the post-content block? IMO, it fits the same pattern really well.

Yeah, I think whatever visibility and notifications we settle on would make sense for post-content as well?  It is the same in that it is a separate entity whose changes will be applied to everywhere it is used.  Perhaps some other blocks as well?  If it gets to a certain point maybe we should consider making some sort of hook or HOC to enable this functionality as opposed to writing it into every block. 🤔   although im not sure if there would be enough blocks needing this behavior to warrant that?

noahtallen commented:
Re the blue border, I think the idea is to have the blue block border be related to the block you are specifically editing. I think when selecting a child block, you are specifically editing that child block rather than the template part, so IMO the template part border should be slightly different when a child block is selected vs when itself is selected. (for example, i think the [designs here](https://github.com/WordPress/gutenberg/issues/22064#issuecomment-667084051) have a gray border when a child block is focused)

Maybe even a lighter blue color to indicate that it's still related to what you're editing, but not the _specific_ thing you're editing? 🤔 

> Yeah, it seems like the block inserter there is picking up the hover state

Oof, that's unfortunate :(

> I wonder if there is an option to disable this behavior.

It almost feels like this isn't a normal notice but just some info about the block being edited 🤔 Maybe we should build it separately rather than using the notice lib?

> I think whatever visibility and notifications we settle on would make sense for post-content as well

+1 for sure. In my mind, the distinction is an "inner block controller" should get this treatment. Maybe not something like the nav block or site title block. Post content and template part would both be examples of inner block controllers though

Addison-Stavlo commented:
> (for example, i think the designs here have a gray border when a child block is focused)

Oh!  I missed that (or misinterpreted it).  Updated!

> Yeah, it seems like the block inserter there is picking up the hover state
> > Oof, that's unfortunate :(

I wonder if we do something like add/remove an `is-hovered` class on mouseover / mouse leave if that would get around this?  The __experimentalBlock wrapper seems to do this:

https://github.com/WordPress/gutenberg/blob/af53147b3adf1951aeac7bdbbf3b7a574d8bc610/packages/block-editor/src/components/block-list/block-wrapper.js#L253

Although im not noticing it working for blocks that use that in the site editor? 🤔 

> It almost feels like this isn't a normal notice but just some info about the block being edited 🤔 Maybe we should build it separately rather than using the notice lib?

Yeah, I think that might make sense.

noahtallen commented:
I think that hover class is supposed to show up. I think some changes in this PR must remove it, because I noticed that it _was_ working in #24394. 👀

noahtallen commented:
btw @Addison-Stavlo I rebased this on `master` and ended up having to squash all of the commits to avoid weird merge conflicts. my hunch is that we were missing a change related to the hover states, because they were showing up on `master`, but didn't show up in this branch. and no changes here should have caused them to go away :)

noahtallen commented:
The `is_hovered` class is showing up now, but it appears to be removed when you hover over a child block

jeyip committed:
Create attached template part notice

The initial implementation of template part notice reused the notice component from the component library and notice logic from core/notices.

After more discussion, we decided to try an alternative, custom implementation where the notice is attached to the template part block itself.

noahtallen commented:
noticed a little bug. when first loading, the focus mode was still applied:

<img width="1768" alt="Screen Shot 2020-08-07 at 2 49 13 PM" src="https://user-images.githubusercontent.com/6265975/89691231-45d18900-d8bd-11ea-8752-1365c4dcd52d.png">

just posting the screenshot here for discussion:

**Template part block focused**
<img width="1768" alt="Screen Shot 2020-08-07 at 2 50 50 PM" src="https://user-images.githubusercontent.com/6265975/89691299-73b6cd80-d8bd-11ea-968e-8214312cb8a0.png">

I'm kinda seeing four different areas of UI which feel as if they could be combined into the same area. in clockwise order:
- block toolbar
- template part name
- information about changes applying in more than just this one place
- the "header" tag (which is always visible even when selecting a child block

noahtallen commented:
#24450 moves the block title to the toolbar, which is a good way to help make things more consistent

Addison-Stavlo commented:
> It almost feels like this isn't a normal notice but just some info about the block being edited 🤔 Maybe we should build it separately rather than using the notice lib?

I have been thinking about this more and am flip flopping a bit on my opinion.  We probably should be using the notices package as much as we can.  If the snackbar notice is not visible enough, we should add options to the notices package to allow for snackbar notices that do not dismiss automatically after X seconds or to increase their visibility in other ways.  Notices are build with accessibility, screen reader, and other concerns in mind.  It helps to utilize the package because as time goes on and notices are improved or iterated upon, our template part notice would receive these upgrades and remain consistent with the rest of the API.  Building specific in-block notices seems like it could create a bit more pain down the line when they need to be updated, especially if we need to allow similar notices for other entity specific blocks.

Another thing to consider is that the snackbar is a tried and true place to show notices.  Attaching them to the top or bottom of a block will become problematic as they won't be visible if the block is at the top or bottom of the page.  I.E. if we scroll down to the footer and start editing the top blocks in it, the notice attached beneath it may remain off the screen and unseen.  Similar issues with large template parts that may have larger height than the editor height (like maybe with post-content blocks later as they may need a similar notice).  While the snackbar will always be visible in the editor.

Addison-Stavlo commented:
> noticed a little bug. when first loading, the focus mode was still applied:

Right, the setting persists beyond reload.  So we need to check for it and turn it off on editor load (in the same place you were experimenting with enabling select mode)

Addison-Stavlo committed:
turn off focusMode on site editor load

jeyip committed:
Reintroduce snackbar notice

jeyip committed:
Position template part label at the top left of the block

jeyip committed:
Remove redundant notification text

jeyip committed:
Only display template part label when a child block is selected
